larpandoracontent/LArControlFlow/TrackDirectionTool.h

```diff
@@ -0,0 +1,1089 @@
+    void GetTrackLength(HitChargeVector &hitChargeVector, float &trackLength);
+
+    void GetAverageQoverWTrackBody(HitChargeVector &hitChargeVector, float &averageChargeTrackBody);
+
+    void GetQoverWRange(HitChargeVector &hitChargeVector, float &QoverWRange);
+
+    void FitHitChargeVector(HitChargeVector &hitChargeVector, TrackDirectionTool::DirectionFitObject &fitResult, int numberHitsToConsider=1000000);
+
+    void FitHitChargeVector(HitChargeVector &hitChargeVector1, HitChargeVector &hitChargeVector2, TrackDirectionTool::DirectionFitObject &fitResult1, TrackDirectionTool::DirectionFitObject &fitResult2, int numberHitsToConsider=1000000);
+
+    void ComputeProbability(DirectionFitObject &fitResult);
+
+    void PerformFits(HitChargeVector &hitChargeVector, HitChargeVector &forwardsFitPoints, HitChargeVector &backwardsFitPoints, int numberHitsToConsider, float &forwardsChiSquared, float &backwardsChiSquared, int &fitStatus1, int &fitStatus2);
+
+    void GetCalorimetricDirection(const pandora::Cluster* pTargetClusterW, DirectionFitObject &finalDirectionFitObject);
+
+    void AddToSlidingFitCache(const pandora::Cluster *const pCluster);
+
+    const TwoDSlidingFitResult &GetCachedSlidingFit(const pandora::Cluster *const pCluster) const;
+
+    static bool SortHitChargeVectorByRL(HitCharge &hitCharge1, HitCharge &hitCharge2);
+    static bool SortHitChargeVectorByChargeOverWidth(HitCharge &hitCharge1, HitCharge &hitCharge2);
+    static bool SortByDistanceToNN(HitCharge &hitCharge1, HitCharge &hitCharge2);
+    void BinHitChargeVector(HitChargeVector &hitChargeVector, HitChargeVector &binnedHitChargeVector);
+
+    double DensityCorrection(double &T, double &M);
+
+    double BetheBloch(double &T, double &M);
+
+    void FillLookupTable(LookupTable &lookupTable, double M);
+
+    double GetEnergyfromLength(LookupTable &lookupTable, double &trackLength);
+
+    double GetLengthfromEnergy(LookupTable &lookupTable, double &currentEnergy);
+
+
+};
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline TrackDirectionTool::HitCharge::HitCharge(const pandora::CaloHit* caloHit, float &longitudinalPosition, float &hitWidth, float &hitCharge, float &uncertainty) :
+    m_intails((hitCharge/hitWidth) <= 1.4 ? true : false),
+    m_calohit(caloHit),
+    m_longitudinalposition(longitudinalPosition),
+    m_hitwidth(hitWidth),
+    m_charge(hitCharge),
+    m_qoverx(hitCharge/hitWidth),
+    m_uncertainty(uncertainty),
+    m_forwardsfitcharge(0.f),
+    m_forwardssigma(0.f),
+    m_forwardsdelta(0.f),
+    m_forwardschisquared(0.f),
+    m_backwardsfitcharge(0.f),
+    m_backwardssigma(0.f),
+    m_backwardsdelta(0.f),
+    m_backwardschisquared(0.f)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline TrackDirectionTool::HitCharge::HitCharge() :
```

Comment:
Why not?

Reply:
The key question is whether or not you want/need to have a version of this object that doesn't actually describe a hit and its associated properties.

It's sometimes necessary to provide default constructors either for technical reasons, or because a default set of values for the object has a clear meaning. In this case it's not clear to me that having a HitCharge object with a null hit and all other values set to zero is something you would want to create, and if it's not, it's better to avoid defining the default constructor, as this prevents people from accidentally creating an object that doesn't describe anything valid.

Reply:
Right okay. Removed.